Sync enclave

RmbRT opened the pull request:
* Made enclave synchronous, so that all commands are handled in one main goroutine.
* Removed lots of code.
* Reworked epoch and block processors.
* Split code into more files, grouped by topic.

RmbRT committed:
tee/prototype: Removed epoch chain.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
tee/prototype: Deleted unused code.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
operator: Adapted operator test.

- Now retrieving balance proofs in every epoch, since transaction and
  deposit epoch have been merged.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
tee/prototype: Fixed typo in comment.

- Also renamed events.go to blockchain_events.go for clarity.

Signed-off-by: Steffen Rattay <steffen@perun.network>

sebastianst commented on tee/prototype/blockchain.go:
```diff
@@ -5,84 +5,142 @@ package prototype
+// chain remains unchanged.
+func (b *blockchain) PushVerify(
+	block *tee.Block,
+	params *tee.Parameters,
```
Since the parameters stay the same over the lifetime of a `blockchain`, we could consider putting a `*tee.Parameters` pointer into the `struct blockchain` so we don't need to pass it every time.

RmbRT replied:
I don't really like state duplication, I already feel bad for having the parameters inside the Enclave and also in Enclave.State. The blockchain doesn't really own the parameters, and I think therefore it shouldn't have them. it only needs the parameters to know the context in which it operates, but the parameters are not really part of what it is or does.

RmbRT replied:
Same with the epoch. It needs it to verify the deposits and exits, but the epoch shouldn't be part of the blockchain's state.

sebastianst commented on tee/prototype/blockchain.go:
```diff
@@ -5,84 +5,142 @@ package prototype
+	// TODO: Extend validation to test consensus (PoW).
+	//   This function then probably becomes a method on nextchain after we add a
+	//   consensus engine to nextchain.
+	switch true {
```
```suggestion
	switch {
```

sebastianst commented on tee/prototype/blockchain.go:
```diff
@@ -5,84 +5,142 @@ package prototype
-	if b.offset >= blockNum {
-		return
+// extractEvents extracts all deposit and exit events from a block.
+func (b *blockchain) extractEvents(
```
If I see correctly, functions `extractEvents` and `filterLogs` don't work on any fields of `blockchain`, so you might as well change them to free functions. Unless you put `params` into the struct... (at least for this method)

For my taste, methods should be methods on the primary object that they operate on. So in this case, either a method on `tee.Block` or if this doesn't make contextually sense, a free function. Note that parameter `epoch` is unused... something is wrong with our linter (or I am blind).

RmbRT replied:
removed epoch, but kept the functions as freestanding for now. I don't want to mess around in the rest of the code base outside the tee/prototype directory, as this would also make me rework of the whole tee.Block API. I'll open an issue for this.

sebastianst commented on tee/prototype/blockchain.go:
```diff
@@ -5,84 +5,142 @@ package prototype
+// filterLogs partitions logs into different buckets of matching predicates.
+// Only logs from the Erdstall contract are filtered and other logs, as well as
+// those without a matching predicate, are discarded.
+func (b *blockchain) filterLogs(logs []*types.Log, preds []logPredicate, contract common.Address) [][]*types.Log {
```
free function

sebastianst commented on tee/prototype/enclave_command.go:
```diff
@@ -0,0 +1,167 @@
+type (
+	// command represents all enclave command types.
+	command interface {
+		isACommand()
```
I'd just name it `command()`.

RmbRT replied:
:+1:

sebastianst commented on tee/prototype/enclave_command.go:
```diff
@@ -0,0 +1,167 @@
+// enclave's state to reflect the new block. On error, the enclave remains
+// unchanged.
+func (e *Enclave) pushBlock(block *tee.Block) (phaseShift bool, _ error) {
+
```
remove \n

sebastianst commented on tee/prototype/enclave_command.go:
```diff
@@ -0,0 +1,167 @@
+
+	phaseEnd := e.IsAtPhaseEnd()
+
+	return phaseEnd, nil
```
```suggestion
	return e.IsAtPhaseEnd(), nil
```

sebastianst replied:
I'd actually just let `pushBlock` return `error` and let the caller check for `e.IsAtPhaseEnd()`.

sebastianst commented on tee/prototype/enclave_command.go:
```diff
@@ -0,0 +1,167 @@
+	// Push block and if there is a phase shift, progress into the next epoch.
+	if phaseShift, err := e.pushBlock(block); err != nil {
+		return err
+	} else if phaseShift {
```
I'd actually just let `pushBlock` return `error` and check here directly for `e.IsAtPhaseEnd()`. Better decouplement of what `pushBlock` is doing and what the caller is doing.

sebastianst commented on tee/prototype/enclave.go:
```diff
@@ -3,211 +3,196 @@
 	"github.com/perun-network/erdstall/tee"
+
+	perrors "perun.network/go-perun/pkg/errors"
+	"perun.network/go-perun/pkg/sync/atomic"
```
go-perun imports should be grouped together with other external imports like go-ethereum. I suggest you setup your IDE to call [`goimports`](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) at file save like `goimports -local github.com/perun-network/erdstall`.

sebastianst commented on tee/prototype/enclave.go:
```diff
@@ -3,211 +3,196 @@
 	Enclave struct {
-		params  tee.Parameters
+		params *tee.Parameters // The enclave's parameters.
+		*State                 // The enclave's essential state.
```
Why as pointer?

RmbRT replied:
So that it is nil if the enclave is not yet fully initialised. It's just as a failsafe so that it panics on invalid usage.

sebastianst commented on tee/prototype/enclave.go:
```diff
@@ -3,211 +3,196 @@
+	log.Info("Enclave: shutting down when phase is finished")
+	select {
+	case e.commands <- &shutdownCmd{}:
+	case <-e.stopped:
```
I think it's fair to panic if `Shutdown` is called twice.

sebastianst commented on tee/prototype/enclave.go:
```diff
@@ -3,211 +3,196 @@
+// enclave simultaneously. Once this number is surpassed, it is no longer
+// possible to ensure a FIFO execution of commands, and excess commands will
+// result in an error.
+const EnclaveMaxCommandQueue = 256 // for good measure.
```
why public?

sebastianst commented on tee/prototype/enclave.go:
```diff
@@ -3,211 +3,196 @@
-		case e.newBlocks <- blockReq{b, err}:
-		case <-e.done:
+		case errs := <-errCh:
+			errg := perrors.NewGatherer()
+			for _, e := range errs {
+				errg.Add(e)
+			}
```
we should add a constructor for this into `go-perun/pkg`...

RmbRT replied:
created issue 456.

sebastianst commented on tee/prototype/epoch.go:
```diff
@@ -3,81 +3,198 @@
+		balances   map[common.Address]*Acc     // Latest balances.
 
-		balances map[common.Address]*Bal
+		epochOutcome chan Outcome // The last epoch's outcome.
```
```suggestion
		outcome chan Outcome // The last epoch's outcome.
```
what other outcome should it be as a field of `Epoch` ;)

sebastianst commented on tee/prototype/epoch.go:
```diff
@@ -3,81 +3,198 @@
+
+		exitLocked map[common.Address]struct{} // Withdrawing at end of epoch.
+		exitReqs   map[common.Address]struct{} // Requested exits.
+		balances   map[common.Address]*Acc     // Latest balances.
```
IMO the field name should be consistent with the type, so
```suggestion
		accs   map[common.Address]*Acc     // Latest account states.
```
or `accounts` or
```suggestion
		balances   map[common.Address]*Balance     // Latest balances.
```

RmbRT replied:
accs

sebastianst commented on tee/prototype/epoch.go:
```diff
@@ -3,81 +3,198 @@
+	}
 )
 
+// NewEpoch creates an epoch manager that starts at the desired deposit epoch.
```
`manager` is a forbidden word ;)

sebastianst replied:
We probably don't even need to have this function public...

sebastianst commented on tee/prototype/epoch.go:
```diff
@@ -3,81 +3,198 @@
+// Outcome has to be called exactly once after each phase shift and contains all
+// accounts that exited within this epoch, as well as all balances at the end of
+// the epoch. It blocks until the epoch in which it was last called ends.
+func (e *Epoch) Outcome() Outcome { return <-e.epochOutcome }
```
Is there any gain in exposing it as
```suggestion
func (e *Epoch) Outcome() <-chan Outcome { return e.epochOutcome }
```
so that, at the calling site, we can `select` on some other event like enclave closing?

RmbRT replied:
no, it is only called internally in the block processor after the epoch has already shifted, but I guess it would make sense to add an assertion that it really exists, and panic instead of deadlocking in case it does not exist. The enclave cannot close at that point because it only after the block processor is done with the epoch in which a shutdown was requested.

RmbRT replied:
I added a log.Panic in case it is called more than once per epoch or before the epoch shift, because that would indicate a serious logic error.

sebastianst commented on tee/prototype/epoch.go:
```diff
@@ -3,81 +3,198 @@
+	} else if sender.Value.Cmp((*big.Int)(tx.Amount)) < 0 {
+		return errors.New("insufficient balance")
+	} else if (*big.Int)(tx.Amount).Sign() < 0 {
+		return errors.New("negative balance")
```
```suggestion
		return errors.New("negative amount")
```

sebastianst commented on tee/prototype/epoch.go:
```diff
@@ -3,81 +3,198 @@
+		e.balances[tx.Recipient] = &Acc{0, new(big.Int).Set((*big.Int)(tx.Amount))}
+	} else {
+		recipient.Value.Add(recipient.Value, (*big.Int)(tx.Amount))
 	}
```
I think there's code/logic duplication here and in `ApplyDeposits` that should be better encapsulated into a method

sebastianst commented on tee/test/generictest.go:
```diff
@@ -159,17 +159,22 @@ func GenericEnclaveTest(t *testing.T, enc tee.Enclave) {
+	seal("exitPhase", 1)
+	bpsExit, err := enc.BalanceProofs()
+	requiree.NoError(err)
+	requiree.Len(bpsExit, 2, "users should still be in the system (frozen)")
```
"frozen" is misleading because we use this term in the case that the operator goes offline and the contract freezes. Do you mean something like "exit locked" here?

RmbRT replied:
yes, good catch

sebastianst commented on tee/test/generictest.go:
```diff
@@ -159,17 +159,22 @@ func GenericEnclaveTest(t *testing.T, enc tee.Enclave) {
+	requiree.Len(bpsExit, 2, "users should still be in the system (frozen)")
+	_, err = enc.DepositProofs()
+	requiree.NoError(err)
+
```
Also add an assert on length 0 for the deposit proofs.

RmbRT committed:
tee/test: Adapted generic test.

- Now waits one more epoch to exit the system to check that accounts are
  removed.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
tee/prototype: Simplified blockchain.

- Threw out fork handling.
- Only tracking the latest block now.
- Adapted blockchain test.
- Now validates block receipts before pushing a block, so that no errors
  can occur after pushing a block.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
tee/prototype: Improved Enclave.

- Now uses one event loop to synchronously execute commands.
- Moved command implementations into command.go.
- Commands are now synchronous and no longer use channels.
- Enclave now only uses a single goroutine.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
tee/prototype: Separated proof generation.

- Now proof generation is in proofs.go.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
tee/prototype: Simplified epoch.

- Merged deposit and transaction epoch.
- Prepared for 2-phase exit protocol.
- Only one epoch object instead of 3 describe the system state now.
- Transaction processing is now part of the epoch.
- Deposits cannot fail in the enclave any more.
- Added GetOutcome(), which at the epoch shift returns all final
  balances and the accounts that exited within an epoch.
- Allowed transactions to non-existent accounts.

Signed-off-by: Steffen Rattay <steffen@perun.network>
